Add support for startup and installed events for web extensions and

xeenon opened the pull request:
#### c3d35ed194578765a4a11e2f2b3d40acea5b6010
<pre>
Add support for startup and installed events for web extensions and
move Web Extension local storage (if base URL changed).
<a href="https://webkit.org/b/248889">https://webkit.org/b/248889</a>
<a href="https://webkit.org/b/248430">https://webkit.org/b/248430</a>
rdar://problem/105337393
rdar://problem/103261455

Reviewed by Brent Fulgham.

Added support for the runtime.onInstalled and runtime.onStartup events, as well as moving
the local storage if the base URL changes.

* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIRuntimeCocoa.mm:
(WebKit::WebExtensionContext::fireRuntimeStartupEventIfNeeded):
(WebKit::WebExtensionContext::fireRuntimeInstalledEventIfNeeded):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::load):
(WebKit::WebExtensionContext::currentState const):
(WebKit::WebExtensionContext::readStateFromStorage):
(WebKit::WebExtensionContext::moveLocalStorageIfNeeded):
(WebKit::WebExtensionContext::loadBackgroundWebViewDuringLoad):
(WebKit::WebExtensionContext::queueStartupAndInstallEventsForExtensionIfNecessary):
(WebKit::WebExtensionContext::loadBackgroundPageListenersFromStorage):
(WebKit::WebExtensionContext::saveBackgroundPageListenersToStorage):
(WebKit::WebExtensionContext::performTasksAfterBackgroundContentLoads):
(WebKit::WebExtensionContext::queueEventToFireAfterBackgroundContentLoads):
(WebKit::WebExtensionContext::loadBackgroundPageListenersVersionNumberFromStorage): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::storageDirectory const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::storageIsPersistent const):
* Source/WebKit/UIProcess/Extensions/WebExtensionController.cpp:
(WebKit::WebExtensionController::WebExtensionController):
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
(WebKit::WebExtensionController::isFreshlyCreated const):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
(WebKit::WebExtensionAPIRuntime::onInstalled):
(WebKit::WebExtensionAPIRuntime::onStartup):
(WebKit::toWebAPI):
(WebKit::WebExtensionContextProxy::dispatchRuntimeInstalledEvent):
(WebKit::WebExtensionContextProxy::dispatchRuntimeStartupEvent):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIRuntime.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIRuntime.idl:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIRuntime.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.h:
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionManager runForTimeInterval:]):

Canonical link: <a href="https://commits.webkit.org/268932@main">https://commits.webkit.org/268932@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/758e0f4391c73f1169a239e8efb07de57f914dab

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21109 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/21480 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/22173 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/22987 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/19633 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21349 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/23/builds/24739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/21671 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/22987 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21333 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/23/builds/24739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/22173 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/23841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/21297 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/23/builds/24739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/22173 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/23841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/23/builds/24739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/22173 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/23841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/19879 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/21671 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/19158 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/22173 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/23448 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2607 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/19733 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIRuntimeCocoa.mm:
```diff
@@ -172,6 +172,28 @@
 
+void WebExtensionContext::fireRuntimeStartupEventIfNeeded()
+{
+    // The background content is assumed to be loaded for this event.
```
This would be a great thing to have an ASSERT about, but maybe there is no convenient flag or predicate to check for yet.

xeenon replied:
Yeah, no good way to check or I would have. This is only called once we get the did finish load callback though.

brentfulgham commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -142,6 +144,22 @@ - (void)webViewWebContentProcessDidTerminate:(WKWebView *)webView
+WebExtensionContext::WebProcessProxySet WebExtensionContext::processes(WebExtensionEventListenerType type, WebExtensionContentWorldType contentWorldType) const
+{
+    auto pagesEntry = m_eventListenerPages.find({ type, contentWorldType });
+    if (pagesEntry == m_eventListenerPages.end())
```
Do we need to check if `pagesEntry`'s `WebPageProxy` is nullptr, since it's a WeakPtr? Or does the container discard entries that have gone nullptr? (Or maybe I misunderstood the type declaration for this container).

xeenon replied:
Yeah `WeakHashCountedSet` discards null entries before access.

brentfulgham commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -142,6 +144,22 @@ - (void)webViewWebContentProcessDidTerminate:(WKWebView *)webView
+
+    WebProcessProxySet result;
+    for (auto entry : pagesEntry->value) {
+        Ref process = entry.key.process();
```
So any active WeakPtr for a page is elevated to a Ref-counted thing to keep it alive. That sounds reasonable, but I wanted to double-check that the goal here is to hold the page "alive" until the extension is finished with it?

brentfulgham replied:
Oh, I see -- this was just code that moved from `WebExtensionContext::processes`. But I do wonder about the ownership change. Seems like it's intentional, so nothing to do.

xeenon replied:
Yeah this is moved code. The `Ref` thing was recently done by @cdumez. We only need to keep it around to send the message. it is fine if it goes away after that or even before a reply is sent.

xeenon replied:
I actually didn't need to move this, I originally had added some ObjC code here but removed it. I will just move it back and no change this.

brentfulgham reviewed (approved):
Looks good. r=me

xeenon committed:
Add support for startup and installed events for web extensions and
move Web Extension local storage (if base URL changed).
https://webkit.org/b/248889
https://webkit.org/b/248430
rdar://problem/105337393
rdar://problem/103261455

Reviewed by Brent Fulgham.

Added support for the runtime.onInstalled and runtime.onStartup events, as well as moving
the local storage if the base URL changes.

* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIRuntimeCocoa.mm:
(WebKit::WebExtensionContext::fireRuntimeStartupEventIfNeeded):
(WebKit::WebExtensionContext::fireRuntimeInstalledEventIfNeeded):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::load):
(WebKit::WebExtensionContext::currentState const):
(WebKit::WebExtensionContext::readStateFromStorage):
(WebKit::WebExtensionContext::moveLocalStorageIfNeeded):
(WebKit::WebExtensionContext::loadBackgroundWebViewDuringLoad):
(WebKit::WebExtensionContext::queueStartupAndInstallEventsForExtensionIfNecessary):
(WebKit::WebExtensionContext::loadBackgroundPageListenersFromStorage):
(WebKit::WebExtensionContext::saveBackgroundPageListenersToStorage):
(WebKit::WebExtensionContext::performTasksAfterBackgroundContentLoads):
(WebKit::WebExtensionContext::queueEventToFireAfterBackgroundContentLoads):
(WebKit::WebExtensionContext::loadBackgroundPageListenersVersionNumberFromStorage): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::storageDirectory const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::storageIsPersistent const):
* Source/WebKit/UIProcess/Extensions/WebExtensionController.cpp:
(WebKit::WebExtensionController::WebExtensionController):
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
(WebKit::WebExtensionController::isFreshlyCreated const):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
(WebKit::WebExtensionAPIRuntime::onInstalled):
(WebKit::WebExtensionAPIRuntime::onStartup):
(WebKit::toWebAPI):
(WebKit::WebExtensionContextProxy::dispatchRuntimeInstalledEvent):
(WebKit::WebExtensionContextProxy::dispatchRuntimeStartupEvent):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIRuntime.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIRuntime.idl:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIRuntime.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.h:
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionManager runForTimeInterval:]):

Canonical link: https://commits.webkit.org/268932@main
